Added preview window

hrueger opened the pull request.

hrueger committed:
added preview window

Envek commented on obsRecorder.js:
```diff
@@ -18,6 +19,23 @@ function initialize(webContents) {
+    displayId,
+  );
+  osn.NodeObs.OBS_content_setShouldDrawUI(displayId, false);
+  osn.NodeObs.OBS_content_resizeDisplay(displayId, displayWidth, displayHeight);
```
In some cases (HiDPI displays and scaled UI) video width and height here will produce video that doesn't fill window completely (or overflows it). Because BrowserWindow is being created with virtual scaled pixels but here we specify hardware pixels.

So scaling logic from `setupSources` should also be applied here.

Envek reviewed (changes requested):
Thank you very much for this Pull Request! It works!

However, I've found some issues with it:

1. Preview window should be destroyed on application exit (before obs studio is shut down) or error occurs (try to close main window)
2. Let's move all the new code into some function?
3. What about using scene produced in `setupSources` in preview window? It will allow us to modify scene and automatically get identical results in both preview window and recorded files
4. See separate comment about scaling issues

hrueger committed:
moved into new function and made preview win child

hrueger committed:
display scene

hrueger committed:
fixed error when shutting down

hrueger commented:
OK, I fixed issue 1 by making the preview window a child window of the main window. That way it is automatically closed once the main window is closed. And I fixed 2 and 3.

However, I wasn't able to get the scaling working, it always cut off some parts (the taskbar) and I didn't want to just add a random value because this can't be the correct way.

Envek commented:
Thank you!

Envek commented:
Cutted pixels was caused by window title bar and borders being accounted in window width and height. Fixed this by specifying `useContentSize` window option. See https://github.com/Envek/obs-studio-node-example/commit/64c2f861e732d27946c523591a6ffa31185b77bf

Envek commented:
Also I moved preview right into main window (with some dirty hacks) in https://github.com/Envek/obs-studio-node-example/commit/5149f39935c5b3c55bc31502e832d8c86da8fd88

@hrueger, can you please check how it works on your machine? That there is no scaling issues or similar?

hrueger commented:
It's working great @Envek ! On a scaled screen it works, too.
I just found two problems:
1. When I move the window from a scaled to a secondary non-scaled screen, the preview overflows the rest of the window. But I don't know how to detect when the scaling changed...
![problem_01](https://user-images.githubusercontent.com/18656830/81558391-e03d3180-938d-11ea-9694-8e022a6c5c18.png)

2. I have some small black bars around my screen in the preview window, as shown here:
![problem_02](https://user-images.githubusercontent.com/18656830/81558446-f8ad4c00-938d-11ea-8d32-de0a9177c681.png)

Envek commented:
Oh, nice, thank you very much for testing it!

 1. Created separate issue about scaling.

 2. > I have some small black bars around my screen in the preview window

    I think that's because preview window is a bit wider than video itself and OBS fills it with some color. I've seen a function to change this background color somewhere in obs-studio-node.

Envek commented:
Fixed in https://github.com/Envek/obs-studio-node-example/commit/30925a19764c777085c23096bc51e670ff6cf5b8
